gusto/recovery.py

```diff
@@ -0,0 +1,318 @@
+                 args={"DG1": (self.v_out, RW),
+                       "CG1": (self.v1, READ),
+                       "DG0": (self.v0, READ),
+                       "COORDS": (self.coords, READ),
+                       "RIGHT": (self.right, READ)},
+                 iterate=ON_TOP)
+
+
+class Recoverer(object):
+    """
+    An object that 'recovers' a field from a low order space
+    (e.g. DG0) into a higher order space (e.g. CG1). This encompasses
+    the process of interpolating first to a the right space before
+    using the :class:`Averager` object, and also automates the
+    boundary recovery process. If no boundary method is specified,
+    this simply performs the action of the :class: `Averager`.
+
+    :arg v_in: the :class:`ufl.Expr` or
+         :class:`.Function` to project. (e.g. a VDG0 function)
+    :arg v_out: :class:`.Function` to put the result in. (e.g. a CG1 function)
+    :arg VDG: optional :class:`.FunctionSpace`. If not None, v_in is interpolated
+         to this space first before recovery happens.
+    :arg boundary_method: a string defining which type of method needs to be
+         used at the boundaries. Valid options are 'density' or 'velocity'.
+    """
+
+    def __init__(self, v_in, v_out, VDG=None, boundary_method=None):
+
+        # check if v_in is valid
+        if isinstance(v_in, expression.Expression) or not isinstance(v_in, (ufl.core.expr.Expr, function.Function)):
+            raise ValueError("Can only recover UFL expression or Functions not '%s'" % type(v_in))
+
+        self.v_in = v_in
+        self.v_out = v_out
+        self.V = v_out.function_space()
+        if VDG is not None:
+            self.v = Function(VDG)
+            self.interpolator = Interpolator(v_in, self.v)
+        else:
+            self.v = v_in
+            self.interpolator = None
+
+        self.VDG = VDG
+        self.boundary_method = boundary_method
+        self.averager = Averager(self.v, self.v_out)
+
+        # check boundary method options are valid
+        if boundary_method is not None:
+            if boundary_method != 'density' and boundary_method != 'velocity':
+                raise ValueError("Specified boundary_method % not valid" % boundary_method)
+            if VDG is None:
+                raise ValueError("If boundary_method is specified, VDG also needs specifying.")
+
+            # now specify things that we'll need if we are doing boundary recovery
+            if boundary_method == 'density':
+                # check dimensions
+                if self.V.value_size != 1:
+                    raise ValueError('This method only works for scalar functions.')
+                self.boundary_recoverer = Boundary_Recoverer(self.v_in, self.v_out, self.v)
+            if boundary_method == 'velocity':
+                # check dimensions
+                if self.V.value_size != 2:
+                    raise ValueError('This method only works for 2D vector functions.')
```

Comment:
Are there plans on extending this to 3D? If so, this should also be a `NotImplementedError`. I know this sounds nitpicky, but `ValueError` implies using this `Recoverer` on a 3D problem is not mathematically sensible when in reality I'm sure this approach extends?

Reply:
Yes!